src/derived/interpolate.js

```diff
@@ -0,0 +1,108 @@
+import {
+  cond,
+  lessThan,
+  multiply,
+  sub,
+  add,
+  divide,
+  greaterThan,
+} from '../base';
+import invariant from 'fbjs/lib/invariant';
+import AnimatedNode from '../core/AnimatedNode';
+
+function interpolateInternalSingle(value, inputRange, outputRange, offset) {
+  const inS = inputRange[offset];
+  const inE = inputRange[offset + 1];
+  const outS = outputRange[offset];
+  const outE = outputRange[offset + 1];
+  const progress = divide(sub(value, inS), sub(inE, inS));
+  return add(outS, multiply(progress, sub(outE, outS)));
+}
+
+function interpolateInternal(value, inputRange, outputRange, offset = 0) {
+  if (inputRange.length - offset === 2) {
+    return interpolateInternalSingle(value, inputRange, outputRange, offset);
+  }
+  return cond(
+    lessThan(value, inputRange[offset + 1]),
+    interpolateInternalSingle(value, inputRange, outputRange, offset),
+    interpolateInternal(value, inputRange, outputRange, offset + 1)
+  );
+}
+
+export const Extrapolate = {
+  EXTEND: 'EXTEND',
+  CLAMP: 'CLAMP',
+  IDENTITY: 'IDENTITY',
+};
+
+function checkIncreasing(name, arr) {
+  for (let i = 1; i < arr.length; ++i) {
+    // We can't validate animated nodes in JS.
+    if (arr[i] instanceof AnimatedNode || arr[i - 1] instanceof AnimatedNode)
+      continue;
+    invariant(
```

Comment:
For invariant statements I prefer to use the following format:
```
invariant(arr[i] >= arr[i-1], '%s must be monotonically increasting (%s)', name, arr);
```

This way we avoid creating error strings when condition is met and only do it when it invariant is ought to throw an error

Reply:
@DylanVann,
Would you mind changing it to "non-decreasing"?

Reply:
@kmagiera I updated the usages of `invariant`. I wasn't aware of that syntax, I like it.

@osdnk I do think `checkIncreasing` is more clear, as it will only pass if it is increasing (`arr[i] >= arr[i-1]`). Also the error message is "must be monotonically increasing", so it makes sense for the check to be called `checkIncreasing` (or `checkMonotonicallyIncreasing`).

Reply:
@DylanVann It's not a big deal, but there's difference between 'non-decreasing' and 'increasing'

https://github.com/facebook/react-native/commit/cb1bdf1e37236f5147ee5ef745573c0ced1b4f14
I did make a change in RN-core related to this issue recently with explanation

Reply:
@osdnk Oh, sorry, you're 100% right. I'll change it.